crates/miden-objects/src/batch/proposed_batch.rs

```diff
@@ -0,0 +1,465 @@
+        // --------------------------------------------------------------------------------------------
+
+        if chain_mmr.chain_length() != block_header.block_num() {
+            return Err(BatchError::InconsistentChainLength {
+                expected: block_header.block_num(),
+                actual: chain_mmr.chain_length(),
+            });
+        }
+
+        let hashed_peaks = chain_mmr.peaks().hash_peaks();
+        if hashed_peaks != block_header.chain_root() {
+            return Err(BatchError::InconsistentChainRoot {
+                expected: block_header.chain_root(),
+                actual: hashed_peaks,
+            });
+        }
+
+        // Verify all block references from the transactions are in the chain.
+        // --------------------------------------------------------------------------------------------
+
+        // Aggregate block references into a set since the chain MMR does not index by hash.
+        let mut block_references =
+            BTreeSet::from_iter(chain_mmr.block_headers_iter().map(BlockHeader::hash));
+        // Insert the block referenced by the batch to consider it authenticated. We can assume this
+        // because the block kernel will verify the block hash as it is a public input to the batch
+        // kernel.
+        block_references.insert(block_header.hash());
+
+        for tx in transactions.iter() {
+            if !block_references.contains(&tx.block_ref()) {
+                return Err(BatchError::MissingTransactionBlockReference {
+                    block_reference: tx.block_ref(),
+                    transaction_id: tx.id(),
+                });
+            }
+        }
+
+        // Compute batch ID.
+        // --------------------------------------------------------------------------------------------
+
+        let id = BatchId::compute(transactions.iter().map(|tx| tx.id()));
+
+        // Aggregate individual tx-level account updates into a batch-level account update - one per
+        // account.
+        // --------------------------------------------------------------------------------------------
+
+        // Populate batch output notes and updated accounts.
+        let mut account_updates = BTreeMap::<AccountId, BatchAccountUpdate>::new();
+        let mut batch_expiration_block_num = BlockNumber::from(u32::MAX);
+        for tx in transactions.iter() {
+            // Merge account updates so that state transitions A->B->C become A->C.
+            match account_updates.entry(tx.account_id()) {
+                Entry::Vacant(vacant) => {
+                    let batch_account_update = BatchAccountUpdate::new(
+                        tx.account_id(),
+                        tx.account_update().init_state_hash(),
+                        tx.account_update().final_state_hash(),
+                        vec![tx.id()],
+                        tx.account_update().details().clone(),
+                    );
```

Comment:
Should we move this logic into something like `BatchAccountUpdate::from_transaction()`?